src/sensitivity.jl

```diff
@@ -1,44 +1,48 @@
 the types specified by `type_tuple`. If none of the arguments are `Node`s then the method
-of `f` specified by `invoke_type_tuple` is invoked.
+of `f` specified by `invoke_type_tuple` is invoked. If applicable, keyword arguments
+should be provided as a `NamedTuple` and be added to the generated function's signature.
 """
-macro union_intercepts(f::Symbol, type_tuple::Expr, invoke_type_tuple::Expr)
-    return esc(union_intercepts(f, type_tuple, invoke_type_tuple))
+macro union_intercepts(f::Symbol, type_tuple::Expr, invoke_type_tuple::Expr, kwargs::Expr=:(()))
+    kwargs.head === :tuple || throw(ArgumentError("malformed keyword argument specification"))
+    return esc(union_intercepts(f, type_tuple, invoke_type_tuple; eval(kwargs)...))
 end
 
 """
     union_intercepts(f::Symbol, type_tuple::Expr, invoke_type_tuple::Expr)
 
 The work-horse for `@union_intercepts`.
 """
-function union_intercepts(f::Symbol, type_tuple::Expr, invoke_type_tuple::Expr)
-    call, arg_names = get_union_call(f, type_tuple)
-    body = get_body(f, type_tuple, arg_names, invoke_type_tuple)
-    return Expr(:macrocall, Symbol("@generated"), Expr(:function, call, body))
+function union_intercepts(f::Symbol, type_tuple::Expr, invoke_type_tuple::Expr; kwargs...)
+    call, arg_names = get_union_call(f, type_tuple; kwargs...)
+    body = get_body(f, type_tuple, arg_names, invoke_type_tuple; kwargs...)
+    return Expr(:macrocall, Symbol("@generated"), nothing, Expr(:function, call, body))
 end
 
 """
-    @explicit_intercepts(f::Symbol, type_tuple::Expr, is_node::Expr)
+    @explicit_intercepts(f::Symbol, type_tuple::Expr, is_node::Expr[, kwargs::Expr])
     @explicit_intercepts(f::Symbol, type_tuple::Expr)
 
 Create a collection of methods which intecept the function calls to `f` in which at least
 one argument is a `Node`. Types of arguments are specified by the type tuple expression
 in `type_tuple`. If there are arguments which are not differentiable, they can be
 specified by providing a boolean vector `is_node` which indicates those arguments that are
-differentiable with `true` values and those which are not as `false`.
+differentiable with `true` values and those which are not as `false`. Keyword arguments
+to add to the function signature can be specified in `kwargs`, which must be a `NamedTuple`.
 """
-macro explicit_intercepts(f::SymOrExpr, type_tuple::Expr, is_node::Expr)
-    return esc(explicit_intercepts(f, type_tuple, eval(is_node)))
-end
-macro explicit_intercepts(f::SymOrExpr, type_tuple::Expr)
-    is_node = [true for _ in get_types(get_body(type_tuple))]
-    return esc(explicit_intercepts(f, type_tuple, is_node))
+macro explicit_intercepts(
+    f::SymOrExpr,
+    type_tuple::Expr,
+    is_node::Expr=:([true for _ in $(get_types(get_body(type_tuple)))]),
+    kwargs::Expr=:(()),
+)
+    return esc(explicit_intercepts(f, type_tuple, eval(is_node); eval(kwargs)...))
```

Comment:
The `eval`s seem problematic (not your fault). Could we possibly do something like:
```julia
is_node_var = gensym("is_node")
kwargs_var = gensym("kwargs")
expr = quote
    local $is_node_var = $is_node
    local $kwargs_var = $kwargs
    Nabla.explicit_intercepts($(QuoteNode(f)), $(QuoteNode(type_tuple)), $is_node_var, $kwargs_var)
end
return esc(expr)
```

Reply:
I can't quite explain why, but doing that seems to cause weird test failures, e.g.
```
core: Error During Test at /home/alex/.julia/dev/Nabla/test/core.jl:114
  Test threw exception MethodError(*, (3.0, Leaf{Float64} 6.0), 0x0000000000006c69)
  Expression: ∇f(6.0) == (3.0,)
  MethodError: no method matching *(::Float64, ::Leaf{Float64})
  Closest candidates are:
    *(::Any, ::Any, !Matched::Any, !Matched::Any...) at operators.jl:502
    *(::Float64, !Matched::Float64) at float.jl:399
    *(::AbstractFloat, !Matched::Bool) at bool.jl:120
    ...
  Stacktrace:
   [1] (::getfield(Main, Symbol("#f#5")))(::Leaf{Float64}) at /home/alex/.julia/dev/Nabla/test/core.jl:108
   [2] (::getfield(Nabla, Symbol("##8#10")){getfield(Main, Symbol("#f#5")),Bool})(::Float64) at /home/alex/.julia/dev/Nabla/src/core.jl:180
   [3] macro expansion at /home/alex/.julia/dev/Nabla/test/core.jl:114 [inlined]
   [4] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v0.7/Test/src/Test.jl:1079 [inlined]
   [5] top-level scope at /home/alex/.julia/dev/Nabla/test/core.jl:3
```

Reply:
I think because `explicit_intercepts` the function returns an `Expr`, which means that the macro now returns an `Expr` of an `Expr`, so things aren't getting evaluated properly.

Reply:
Would the `eval`s be less problematic if I first ensure that they represent the proper expression types, so someone can't put in like `@explicit_intercepts sin Tuple{Int} rm("/", force=true, recursive=true)`?

Reply:
It seems like this is more work than I would have hoped. Maybe open an issue about eliminating the `eval`s later and we can call it a day :)

Reply:
Yeah, I spent a while on that specifically the other day but decided to just leave it and hope nobody noticed. :stuck_out_tongue: Should I add the expression validation to at least make it a bit safer, or no?

Reply:
I'm pro leaving this be and moving on. I will be re-writing quite a lot of this code in light of Cassette's existence at some point in the not too distant future, so this code could well change.

Reply:
Sounds good. I'll call this resolved then.